core/server/common/src/main/java/alluxio/master/StateLockManager.java

```diff
@@ -158,12 +167,36 @@ public LockResource lockShared() throws InterruptedException {
     }
     // Register thread for interrupt cycle.
     mSharedWaitersAndHolders.add(Thread.currentThread());
-    // Grab the lock interruptibly.
-    mStateLock.readLock().lockInterruptibly();
+    String threadName = Thread.currentThread().getName();
+    mSharedLockHolders.computeIfAbsent(threadName, k -> new LongAdder()).increment();
+    if (mSharedLockHolders.get(threadName).longValue() > mLogThreshold) {
+      Exception e = new Exception("Thread recursion is deeper than " + mLogThreshold);
+      LOG.warn("Current thread is {}. All state lock holders are {}",
+          threadName, mSharedLockHolders, e);
+    }
+    try {
+      // Grab the lock interruptibly.
+      mStateLock.readLock().lockInterruptibly();
+    } catch (Error e) {
+      // An Error is thrown when the lock is acquired 65536 times, log the jstack before exiting
+      LOG.error("Logging all thread stacks before exiting", e);
+      ThreadUtils.logAllThreads();
```

Comment:
it's almost impossible for the support to get in and take jstack when this issue happens, so i'm just dumping before exit. In fact I figure this should be the best approach for many similar situations. FYI @yuzhu

Reply:
should we do this for this "65536" issue alone?

Reply:
Yes the 65536 issue throws `Error`
```
2023-02-07 01:39:33,379 ERROR ProcessUtils - Uncaught exception while running Alluxio master @sj1hdp485.corp.adobe.com:19998, stopping it and exiting. Exception "java.lang.Error: Maximum lock count exceeded", Root Cause "java.lang.Error: Maximum lock count exceeded"
java.lang.Error: Maximum lock count exceeded
	at java.base/java.util.concurrent.locks.ReentrantReadWriteLock$Sync.fullTryAcquireShared(ReentrantReadWriteLock.java:535)
	at java.base/java.util.concurrent.locks.ReentrantReadWriteLock$Sync.tryAcquireShared(ReentrantReadWriteLock.java:494)
	at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1344)
	at java.base/java.util.concurrent.locks.ReentrantReadWriteLock$ReadLock.lockInterruptibly(ReentrantReadWriteLock.java:783)
	at alluxio.master.StateLockManager.lockShared(StateLockManager.java:162)
	at alluxio.master.AbstractMaster.createJournalContext(AbstractMaster.java:141)
	at ...
```